### PR TITLE
8293618: x86: Wrong code generation in class Assembler

### PR DESCRIPTION
**Pull request by merykitty**
Hi,

This patch fixes some issues in the code generation of x86 assembler:

- `Assembler::testl` misses `prefix(dst)`
- `Assembler::addw` misses the 0x66 prefix
- `Assembler::emit_operand` needs the length of the instruction from the address operand, this is often forgotten, making this parameter explicit to prevent potential issues
- The assembler should not do optimisations that change the actual emitted instructions, these should be moved to `MacroAssembler` instead

AFAICT there is no failure due to these mistakes. Please take a look and give reviews.
Thanks you very much.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293618](https://bugs.openjdk.org/browse/JDK-8293618): x86: Wrong code generation in class Assembler

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10240/head:pull/10240` \
`$ git checkout pull/10240`

Update a local copy of the PR: \
`$ git checkout pull/10240` \
`$ git pull https://git.openjdk.org/jdk pull/10240/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10240`

View PR using the GUI difftool: \
`$ git pr show -t 10240`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10240.diff">https://git.openjdk.org/jdk/pull/10240.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — COMMENTED** [review]
Seems reasonable. I will test it.

**Review by vnkozlov — APPROVED** [review]
Testing passed.

**Comment by merykitty** [comment]
@vnkozlov Thanks a lot for your testing.

File: src/hotspot/cpu/x86/assembler_x86.cpp [hunk 1]
```diff
@@ -348,6 +348,12 @@ void Assembler::emit_arith_operand(int op1, Register rm, Address adr, int32_t im
   }
 }
 
+void Assembler::emit_arith_operand_imm32(int op1, Register rm, Address adr, int32_t imm32) {
```
Comment (TobiHartmann) [inline src/hotspot/cpu/x86/assembler_x86.cpp:351]: Why is this needed?

Comment (merykitty) [reply src/hotspot/cpu/x86/assembler_x86.cpp:351]: @TobiHartmann Thanks for taking a look, method entry barrier needs a 4-byte immediate so it can be patched. This method is used in `Assembler::cmpl_imm32`

Comment (TobiHartmann) [reply src/hotspot/cpu/x86/assembler_x86.cpp:351]: Okay, so code before emitted a 8-bit immediate. Why didn't that cause any issues when patching?

Comment (merykitty) [reply src/hotspot/cpu/x86/assembler_x86.cpp:351]: `cmpl(Address, int)` unconditionally emits a 32-bit immediate before, I make it emit 8-bit immediate where possible and create `cmpl_imm32` to strictly emit a 32-bit immediate, similar to how we have `subl_imm32` and `subq_imm32`.

Comment (TobiHartmann) [reply src/hotspot/cpu/x86/assembler_x86.cpp:351]: Ah! Now I got it. Thanks for the explanation.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by merykitty** [comment]
@vnkozlov @TobiHartmann Thanks a lot for your reviews.
/integrate
